src/Cech_complex/include/gudhi/Cech_complex_blocker.h

```diff
@@ -30,33 +31,111 @@ namespace cech_complex {
+    return std::make_pair(std::move(c), std::move(r));
+  }
+
  public:
+
   /** \internal \brief Čech complex blocker operator() - the oracle - assigns the filtration value from the simplex
    * radius and returns if the simplex expansion must be blocked.
    *  \param[in] sh The Simplex_handle.
    *  \return true if the simplex radius is greater than the Cech_complex max_radius*/
   bool operator()(Simplex_handle sh) {
-    Point_cloud points;
-    for (auto vertex : sc_ptr_->simplex_vertex_range(sh)) {
-      points.push_back(cc_ptr_->get_point(vertex));
+    using Point_cloud =  std::vector<Point_d>;
+    CGAL::NT_converter<FT, Filtration_value> cast_to_fv;
+    Filtration_value radius = 0;
+    bool is_min_enclos_ball = false;
+
+    // for each face of simplex sh, test outsider point is indeed inside enclosing ball, if yes, take it and exit loop, otherwise, new sphere is circumsphere of all vertices
+    for (auto face : sc_ptr_->boundary_simplex_range(sh)) {
+        // Find which vertex of sh is missing in face. We rely on the fact that simplex_vertex_range is sorted.
+        auto longlist = sc_ptr_->simplex_vertex_range(sh);
+        auto shortlist = sc_ptr_->simplex_vertex_range(face);
+
+        auto longiter = std::begin(longlist);
+        auto shortiter = std::begin(shortlist);
+        auto enditer = std::end(shortlist);
+        while(shortiter != enditer && *longiter == *shortiter) { ++longiter; ++shortiter; }
+        auto extra = *longiter; // Vertex_handle
+
+        Sphere sph;
+        auto k = sc_ptr_->key(face);
+        if(k != sc_ptr_->null_key()) {
+            sph = cc_ptr_->get_cache().at(k);
+        }
+        else {
+            Point_cloud face_points;
+            for (auto vertex : sc_ptr_->simplex_vertex_range(face)) {
+                face_points.push_back(cc_ptr_->get_point(vertex));
+#ifdef DEBUG_TRACES
+                std::clog << "#(" << vertex << ")#";
+#endif  // DEBUG_TRACES
+            }
+            sph = get_sphere(face_points.cbegin(), face_points.cend());
+            // Put edge sphere in cache
+            sc_ptr_->assign_key(face, cc_ptr_->get_cache().size());
+            cc_ptr_->get_cache().push_back(sph);
+            // Clear face_points
+            face_points.clear();
+        }
+        // Check if the minimal enclosing ball of current face contains the extra point
+        if (kernel_.squared_distance_d_object()(sph.first, cc_ptr_->get_point(extra)) <= sph.second) {
 #ifdef DEBUG_TRACES
-      std::clog << "#(" << vertex << ")#";
+            std::clog << "center: " << sph.first << ", radius: " <<  radius << std::endl;
 #endif  // DEBUG_TRACES
+            is_min_enclos_ball = true;
+#if CGAL_VERSION_NR >= 1050000000
+            if(exact_) CGAL::exact(sph.second);
+#endif
+            radius = std::sqrt(cast_to_fv(sph.second));
+            sc_ptr_->assign_key(sh, cc_ptr_->get_cache().size());
+            cc_ptr_->get_cache().push_back(sph);
```

Comment:
(for later) Maybe we could assign to sh the same key as the face instead of inserting a second copy of the sphere in the cache.

Reply:
cf. #650